### PR TITLE
Example metrics server using Hyper

### PR DESCRIPTION
**Pull request by adamchalmers**
Say your Rust app is not a web server. You probably don't need to pull in a whole web framework like Axum or Actix-web just to serve one little OpenMetrics endpoint. You can just use Hyper instead. Here's an example of how to do that.

I also added this an example target in Cargo.toml, so that you can run it with cargo:

```sh
$ cargo run --example hyper  &
$ curl localhost:8001
# HELP tokio_hyper_example_example_counter An example counter that will get incremented automatically.
# TYPE tokio_hyper_example_example_counter counter
tokio_hyper_example_example_counter_total 6
# EOF
```

### CODE REVIEW

File: examples/hyper.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    move |_req: Request<Body>| {
+        let reg = registry.clone();
+        Box::pin(async move {
+            let mut buf = Vec::new();
+            encode(&mut buf, &reg.clone()).map(|_| {
+                let body = std::str::from_utf8(buf.as_slice()).unwrap().to_string();
+                Response::builder()
+                    .header(
+                        hyper::header::CONTENT_TYPE,
+                        "application/openmetrics-text; version=1.0.0; charset=utf-8",
```
Comment (mxinden) [inline examples/hyper.rs:82]: :pray:

File: examples/hyper.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    net::{IpAddr, Ipv4Addr, SocketAddr},
+    pin::Pin,
+    sync::Arc,
+    time::Duration,
+};
+use tokio::signal::unix::{signal, SignalKind};
+
+lazy_static::lazy_static! {
+    pub static ref EXAMPLE_COUNTER: Counter = Counter::default();
+}
```
Comment (mxinden) [inline examples/hyper.rs:18]: I would like to encourage users to use local metrics, thus I would prefer this example to use a locally instantiated metric.

Would you mind changing this example to use a non-global metric instead?

I am fine with having an example using global metrics in this repository, though I would like that to be a dedicated example, documenting the trade offs a user buys into.

See also https://github.com/prometheus/client_rust/issues/25

File: examples/hyper.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    let registry = register_metrics();
+
+    // Spawn a server to serve the OpenMetrics endpoint.
+    let metrics_addr = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 8001);
+    tokio::task::spawn(start_server(metrics_addr, registry));
+
+    loop {
+        EXAMPLE_COUNTER.inc();
+        tokio::time::sleep(Duration::from_secs(1)).await;
+    }
```
Comment (mxinden) [inline examples/hyper.rs:31]: This is a neat idea, though for the sake of consistency with the other examples, I would prefer showcasing a standard HTTP request counter in this example.

See tide example:

https://github.com/prometheus/client_rust/blob/12c1de56116d803c709ed6584c263353727221fa/examples/tide.rs#L47-L57

I think an example should showcase a single thing only. (Though it is hard where the boundaries of a thing are.) I am fine for something like the above to be shown in a separate example.

File: Cargo.toml [hunk 4]
```diff
@@ -34,10 +34,17 @@ quickcheck = "1"
...
+hyper = { version = "0.14.16", features = ["server", "http1", "tcp"] }
+lazy_static = "1.4"
 
 [build-dependencies]
 prost-build = { version = "0.9.0", optional = true }
 
+[[example]]
+name = "hyper"
+crate-type = ["bin"]
+
```
Comment (mxinden) [inline Cargo.toml:47]: Sorry for the potentialy stupid question: Why is this entry needed? I can already do `cargo run --example tide` today. Wouldn't `cargo run --example hyper` work just fine without the lines above?

**Review by mxinden — COMMENTED** [review]
:rocket: Thanks for the help. Very much appreciated.

File: Cargo.toml (hunk 4 above)
Comment (adamchalmers) [reply Cargo.toml:47]: You're right, it's not needed, I mistakenly thought it was.

File: examples/hyper.rs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+pub fn make_handler(
+    registry: Arc<Registry>,
+) -> impl Fn(Request<Body>) -> Pin<Box<dyn Future<Output = io::Result<Response<Body>>> + Send>> {
+    // This closure accepts a request and responds with the OpenMetrics encoding of our metrics.
+    move |_req: Request<Body>| {
+        let reg = registry.clone();
+        Box::pin(async move {
+            let mut buf = Vec::new();
+            encode(&mut buf, &reg.clone()).map(|_| {
+                let body = std::str::from_utf8(buf.as_slice()).unwrap().to_string();
```
Comment (mxinden) [inline examples/hyper.rs:70]: Given that one can build a `Body` from a `[u8]` why is the conversion to a string necessary here?

https://docs.rs/hyper/latest/hyper/struct.Body.html#impl-From%3C%26%27static%20%5Bu8%5D%3E-for-Body

**Review by mxinden — COMMENTED** [review]
Just two comments. Otherwise looks good to me. Thanks for the follow-ups.

File: examples/hyper.rs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    pin::Pin,
+    sync::Arc,
+};
+use tokio::signal::unix::{signal, SignalKind};
+
+#[tokio::main]
+async fn main() {
+    let request_counter = Counter::default();
+
+    let registry = register_metrics(&request_counter);
```
[the remark this answers is not in the document]
Comment (adamchalmers) [reply examples/hyper.rs]: I removed the function, you're right

File: examples/hyper.rs (hunk 5 above)
Comment (adamchalmers) [reply examples/hyper.rs]: Ah great point, thanks for pointing that out. Simplified it.

**Review by mxinden — APPROVED** [review]
Thanks for the help. Thanks for making it easier for others to get started.

**Pull request merged**